src/operator/numpy/np_matrix_op-inl.h

```diff
@@ -784,6 +784,167 @@ inline void HSplitOpBackward(const nnvm::NodeAttrs &attrs,
   }
   SplitOpBackwardImpl<xpu>(attrs, ctx, inputs, req, outputs, real_axis);
 }
+
+struct NumpyAppendParam : public dmlc::Parameter<NumpyAppendParam> {
+  dmlc::optional<int> axis;
+  DMLC_DECLARE_PARAMETER(NumpyAppendParam) {
+    DMLC_DECLARE_FIELD(axis)
+    .set_default(dmlc::optional<int>())
+    .describe("The axis along which `values` are appended.  If `axis` is not"
+              "given, both `arr` and `values` are flattened before use.");
+  }
+};
+
+template<typename xpu>
+void NumpyAppendForward(const nnvm::NodeAttrs& attrs,
+                        const OpContext& ctx,
+                        const std::vector<TBlob>& inputs,
+                        const std::vector<OpReqType>& req,
+                        const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mshadow_op;
+
+  const NumpyAppendParam& param = nnvm::get<NumpyAppendParam>(attrs.parsed);
+  CHECK_EQ(inputs.size(), 2U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+
+  std::vector<TBlob> data;
+  bool axis_is_none = false;
+  for (int i = 0; i < 2; i++) {
+    if (!param.axis.has_value()) {
+      data.push_back(inputs[i].reshape(Shape1(inputs[i].shape_.Size())));
+      axis_is_none = true;
+    } else {
+      data.push_back(inputs[i]);
+    }
+  }
+
+  ConcatParam cparam;
+  cparam.num_args = 2;
+  cparam.dim = axis_is_none ? 0 : param.axis.value();
+  MSHADOW_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+    ConcatOp<xpu, DType> op;
+    op.Init(cparam);
+    op.Forward(ctx, data, req, outputs);
+  });
+}
+
+template<typename xpu>
+void NumpyAppendBackward(const nnvm::NodeAttrs& attrs,
+                         const OpContext& ctx,
+                         const std::vector<TBlob>& inputs,
+                         const std::vector<OpReqType>& req,
+                         const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mshadow_op;
+
+  const NumpyAppendParam& param = nnvm::get<NumpyAppendParam>(attrs.parsed);
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 2U);
+  CHECK_EQ(req.size(), 2U);
+
+  std::vector<TBlob> data;
+  bool axis_is_none = false;
+  for (int i = 0; i < 2; i++) {
+    if (!param.axis.has_value()) {
+      data.push_back(outputs[i].reshape(Shape1(outputs[i].shape_.Size())));
+      axis_is_none = true;
+    } else {
+      data.push_back(outputs[i]);
+    }
```

Comment:
so that you could do
```c++
    if (axis_is_none) {
      data.push_back(outputs[i].reshape(Shape1(outputs[i].shape_.Size())));
    } else {
      data.push_back(outputs[i]);
    }
```